src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -986,24 +980,30 @@ impl LinkCollector<'_, '_> {
                 // To work around this, remove it and resolve relative to the crate root instead.
                 // HACK(jynelson)(2): If we just strip `crate::` then suddenly primitives become ambiguous
                 // (consider `crate::char`). Instead, change it to `self::`. This works because 'self' is now the crate root.
-                resolved_self = format!("self::{}", &path_str["crate::".len()..]);
-                path_str = &resolved_self;
+                path_str = format!("self::{}", &path_str["crate::".len()..]);
                 module_id = DefId { krate: item.def_id.krate, index: CRATE_DEF_INDEX };
             }
 
             match self.resolve_with_disambiguator_cached(
                 disambiguator,
-                item,
-                dox,
-                path_str,
+                &path_str,
                 current_item,
                 module_id,
                 extra_fragment,
-                &ori_link,
-                link_range.clone(),
             ) {
-                Some(x) => x,
-                None => return,
+                Ok(res) => res,
+                Err(err) => {
+                    self.report_resolve_error(
```

Comment:
This piece of code makes it possible for the cached and uncached `resolve_with_disambiguator` to have the same (except for a `mut`) signature. This was the simplest way to solve the result error caching that I could think of.